File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    phone_book.sort((a,b)=>a.length-b.length);
    let map = {};
    for(const num of phone_book){
        let tmp = "";
        for(const x of [...num]){
            tmp+=x;
            if(map[tmp]) return false;
        }
        map[num]=true;
    }
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
class Solution {
    public boolean solution(String[] phone_book) {
        Arrays.sort(phone_book,(a,b)-> a.length()-b.length());
        
        Set<String> set = new HashSet<>();
        
        for(String number : phone_book) {
            String s = "";
            for(int i = 0; i < number.length(); i++) {
                s += number.charAt(i);
                if(set.contains(s)) {
                return false;
                }
            }
            set.add(number);
        }
        
        return true;
    }
}